File: sdk-py/primordia_sdk/dbp.py
"""
P7: DBP - Default/Bankruptcy Primitive
Deterministic agent default and liquidation
"""

from dataclasses import dataclass
from typing import List, Optional, Literal
import time

from .canonical import canonicalize_bytes
from .crypto import hash_bytes, sign, verify


DeclarationType = Literal["VOLUNTARY", "INVOLUNTARY", "AUTOMATIC"]
TriggerType = Literal["MISSED_FC", "NEGATIVE_MBS", "MARGIN_CALL", "TIMEOUT"]
LiquidationMethod = Literal["PRO_RATA", "PRIORITY", "AUCTION"]


@dataclass
class Creditor:
    agent_id: str
    amount_micros: int
    priority: int
    collateralized: bool


@dataclass
class Asset:
    asset_type: str
    value_micros: int
    liquid: bool


@dataclass
class Distribution:
    creditor_id: str
    receives_micros: int
    recovery_bps: int


@dataclass
class Trigger:
    type: TriggerType
    reference_id: str
    trigger_timestamp_ms: int


@dataclass
class ObligationsSnapshot:
    total_owed_micros: int
    creditors: List[Creditor]


@dataclass
class AssetsSnapshot:
    total_value_micros: int
    assets: List[Asset]


@dataclass
class LiquidationPlan:
    method: LiquidationMethod
    distributions: List[Distribution]


@dataclass
class DBP:
    dbp_version: str
    default_id: str
    defaulting_agent_id: str
    declaration_type: DeclarationType
    trigger: Trigger
    obligations_snapshot: ObligationsSnapshot
    assets_snapshot: AssetsSnapshot
    recovery_rate_bps: int
    liquidation_plan: LiquidationPlan
    timestamp_ms: int
    arbiter_agent_id: str
    dbp_hash: str
    signature_ed25519: str

    def to_dict(self) -> dict:
        return {
            "dbp_version": self.dbp_version,
            "default_id": self.default_id,
            "defaulting_agent_id": self.defaulting_agent_id,
            "declaration_type": self.declaration_type,
            "trigger": {
                "type": self.trigger.type,
                "reference_id": self.trigger.reference_id,
                "trigger_timestamp_ms": self.trigger.trigger_timestamp_ms,
            },
            "obligations_snapshot": {
                "total_owed_micros": self.obligations_snapshot.total_owed_micros,
                "creditors": [
                    {"agent_id": c.agent_id, "amount_micros": c.amount_micros,
                     "priority": c.priority, "collateralized": c.collateralized}
                    for c in sorted(self.obligations_snapshot.creditors, key=lambda x: x.agent_id)
                ],
            },
            "assets_snapshot": {
                "total_value_micros": self.assets_snapshot.total_value_micros,
                "assets": [
                    {"asset_type": a.asset_type, "value_micros": a.value_micros, "liquid": a.liquid}
                    for a in sorted(self.assets_snapshot.assets, key=lambda x: x.asset_type)
                ],
            },
            "recovery_rate_bps": self.recovery_rate_bps,
            "liquidation_plan": {
                "method": self.liquidation_plan.method,
                "distributions": [
                    {"creditor_id": d.creditor_id, "receives_micros": d.receives_micros,
                     "recovery_bps": d.recovery_bps}
                    for d in sorted(self.liquidation_plan.distributions, key=lambda x: x.creditor_id)
                ],
            },
            "timestamp_ms": self.timestamp_ms,
            "arbiter_agent_id": self.arbiter_agent_id,
            "dbp_hash": self.dbp_hash,
            "signature_ed25519": self.signature_ed25519,
        }


def compute_distributions(
    creditors: List[Creditor],
    total_assets: int,
    method: LiquidationMethod
) -> List[Distribution]:
    """Compute liquidation distributions based on method."""
    # Handle empty creditors
    if not creditors:
        return []

    total_owed = sum(c.amount_micros for c in creditors)

    # Handle zero total owed
    if total_owed == 0:
        return [
            Distribution(creditor_id=c.agent_id, receives_micros=0, recovery_bps=0)
            for c in creditors
        ]

    if method == "PRO_RATA":
        result = []
        for c in creditors:
            receives = int((c.amount_micros / total_owed) * total_assets)
            recovery = int((receives / c.amount_micros) * 10000) if c.amount_micros > 0 else 0
            result.append(Distribution(
                creditor_id=c.agent_id,
                receives_micros=receives,
                recovery_bps=recovery
            ))
        return result

    if method == "PRIORITY":
        distributions = []
        remaining = total_assets
        sorted_creditors = sorted(creditors, key=lambda x: x.priority)

        for c in sorted_creditors:
            receives = min(c.amount_micros, remaining)
            remaining -= receives
            distributions.append(Distribution(
                creditor_id=c.agent_id,
                receives_micros=receives,
                recovery_bps=int((receives / c.amount_micros) * 10000) if c.amount_micros > 0 else 0
            ))
        return distributions

    # AUCTION - same as PRO_RATA for now
    return compute_distributions(creditors, total_assets, "PRO_RATA")


def make_dbp(
    defaulting_agent_id: str,
    declaration_type: DeclarationType,
    trigger_type: TriggerType,
    trigger_reference_id: str,
    creditors: List[Creditor],
    assets: List[Asset],
    liquidation_method: LiquidationMethod,
    arbiter_agent_id: str,
    arbiter_private_key: str
) -> DBP:
    """Create a signed DBP."""
    now = int(time.time() * 1000)
    total_owed = sum(c.amount_micros for c in creditors)
    total_assets = sum(a.value_micros for a in assets)

    distributions = compute_distributions(creditors, total_assets, liquidation_method)
    total_distributed = sum(d.receives_micros for d in distributions)
    recovery_rate_bps = int((total_distributed / total_owed) * 10000) if total_owed > 0 else 0

    # Content for hashing (excludes default_id, dbp_hash, signature)
    dbp_content = {
        "dbp_version": "0.1",
        "defaulting_agent_id": defaulting_agent_id,
        "declaration_type": declaration_type,
        "trigger": {
            "type": trigger_type,
            "reference_id": trigger_reference_id,
            "trigger_timestamp_ms": now,
        },
        "obligations_snapshot": {
            "total_owed_micros": total_owed,
            "creditors": [
                {"agent_id": c.agent_id, "amount_micros": c.amount_micros,
                 "priority": c.priority, "collateralized": c.collateralized}
                for c in sorted(creditors, key=lambda x: x.agent_id)
            ],
        },
        "assets_snapshot": {
            "total_value_micros": total_assets,
            "assets": [
                {"asset_type": a.asset_type, "value_micros": a.value_micros, "liquid": a.liquid}
                for a in sorted(assets, key=lambda x: x.asset_type)
            ],
        },
        "recovery_rate_bps": recovery_rate_bps,
        "liquidation_plan": {
            "method": liquidation_method,
            "distributions": [
                {"creditor_id": d.creditor_id, "receives_micros": d.receives_micros,
                 "recovery_bps": d.recovery_bps}
                for d in sorted(distributions, key=lambda x: x.creditor_id)
            ],
        },
        "timestamp_ms": now,
        "arbiter_agent_id": arbiter_agent_id,
    }

    # Compute hash from content only
    content_hash = hash_bytes(canonicalize_bytes(dbp_content))

    # Sign
    signature = sign(content_hash, arbiter_private_key)

    return DBP(
        dbp_version="0.1",
        default_id=content_hash,
        defaulting_agent_id=defaulting_agent_id,
        declaration_type=declaration_type,
        trigger=Trigger(
            type=trigger_type,
            reference_id=trigger_reference_id,
            trigger_timestamp_ms=now
        ),
        obligations_snapshot=ObligationsSnapshot(
            total_owed_micros=total_owed,
            creditors=sorted(creditors, key=lambda x: x.agent_id)
        ),
        assets_snapshot=AssetsSnapshot(
            total_value_micros=total_assets,
            assets=sorted(assets, key=lambda x: x.asset_type)
        ),
        recovery_rate_bps=recovery_rate_bps,
        liquidation_plan=LiquidationPlan(
            method=liquidation_method,
            distributions=sorted(distributions, key=lambda x: x.creditor_id)
        ),
        timestamp_ms=now,
        arbiter_agent_id=arbiter_agent_id,
        dbp_hash=content_hash,
        signature_ed25519=signature
    )


def get_dbp_content(dbp: DBP) -> dict:
    """Extract hashable content from DBP (excludes default_id, dbp_hash, signature)."""
    dbp_dict = dbp.to_dict()
    del dbp_dict["default_id"]
    del dbp_dict["dbp_hash"]
    del dbp_dict["signature_ed25519"]
    return dbp_dict


def verify_dbp(dbp: DBP) -> bool:
    """Verify DBP signature."""
    content = get_dbp_content(dbp)
    computed_hash = hash_bytes(canonicalize_bytes(content))
    return verify(computed_hash, dbp.signature_ed25519, dbp.arbiter_agent_id)


def should_auto_default(runway_seconds: float, threshold_seconds: float = 0) -> bool:
    """Check if agent should trigger automatic default."""
    return runway_seconds < threshold_seconds


def trigger_default(
    defaulting_agent_id: str,
    declaration_type: DeclarationType,
    trigger_type: TriggerType,
    trigger_reference_id: str,
    creditors: List[Creditor],
    assets: List[Asset],
    liquidation_method: LiquidationMethod,
    arbiter_agent_id: str,
    arbiter_private_key: str
) -> DBP:
    """
    Trigger a default event for an agent.

    This is a convenience wrapper around make_dbp() with clearer semantics
    for initiating a default/bankruptcy event.

    Args:
        defaulting_agent_id: Agent that is defaulting
        declaration_type: Type of default (VOLUNTARY, INVOLUNTARY, AUTOMATIC)
        trigger_type: What triggered the default (MISSED_FC, NEGATIVE_MBS, MARGIN_CALL, TIMEOUT)
        trigger_reference_id: Reference to the event that triggered default
        creditors: List of creditors with amounts owed
        assets: List of assets available for liquidation
        liquidation_method: How to distribute assets (PRO_RATA, PRIORITY, AUCTION)
        arbiter_agent_id: Agent managing the default process
        arbiter_private_key: Private key for signing the DBP

    Returns:
        Signed DBP object
    """
    return make_dbp(
        defaulting_agent_id=defaulting_agent_id,
        declaration_type=declaration_type,
        trigger_type=trigger_type,
        trigger_reference_id=trigger_reference_id,
        creditors=creditors,
        assets=assets,
        liquidation_method=liquidation_method,
        arbiter_agent_id=arbiter_agent_id,
        arbiter_private_key=arbiter_private_key
    )


def resolve_default(
    dbp: DBP,
    arbiter_public_key: str
) -> dict:
    """
    Resolve a default by verifying and extracting the liquidation plan.

    Args:
        dbp: DBP object to resolve
        arbiter_public_key: Public key of the arbiter to verify signature

    Returns:
        Dictionary containing:
        - valid: bool - whether the DBP is valid
        - distributions: List[Distribution] - liquidation distributions if valid
        - recovery_rate_bps: int - overall recovery rate in basis points
        - error: Optional[str] - error message if invalid
    """
    # Verify the DBP
    content = get_dbp_content(dbp)
    computed_hash = hash_bytes(canonicalize_bytes(content))

    is_valid = verify(computed_hash, dbp.signature_ed25519, arbiter_public_key)

    if not is_valid:
        return {
            "valid": False,
            "distributions": [],
            "recovery_rate_bps": 0,
            "error": "Invalid arbiter signature"
        }

    # Validate the distributions match the creditors
    total_owed = dbp.obligations_snapshot.total_owed_micros
    total_distributed = sum(d.receives_micros for d in dbp.liquidation_plan.distributions)
    total_assets = dbp.assets_snapshot.total_value_micros

    # Verify distributions don't exceed assets
    if total_distributed > total_assets:
        return {
            "valid": False,
            "distributions": [],
            "recovery_rate_bps": 0,
            "error": "Distributions exceed available assets"
        }

    # Verify recovery rate calculation
    expected_recovery = int((total_distributed / total_owed) * 10000) if total_owed > 0 else 0
    if dbp.recovery_rate_bps != expected_recovery:
        return {
            "valid": False,
            "distributions": [],
            "recovery_rate_bps": 0,
            "error": f"Recovery rate mismatch: expected {expected_recovery}, got {dbp.recovery_rate_bps}"
        }

    return {
        "valid": True,
        "distributions": dbp.liquidation_plan.distributions,
        "recovery_rate_bps": dbp.recovery_rate_bps,
        "error": None
    }

File: sdk-ts/src/dbp.ts
/**
 * P7: DBP - Default/Bankruptcy Primitive
 * Deterministic agent default and liquidation
 */

import { canonicalize, canonicalizeBytes } from './canonical.js';
import { hash, sign, verify } from './crypto.js';

export type DeclarationType = 'VOLUNTARY' | 'INVOLUNTARY' | 'AUTOMATIC';
export type TriggerType = 'MISSED_FC' | 'NEGATIVE_MBS' | 'MARGIN_CALL' | 'TIMEOUT';
export type LiquidationMethod = 'PRO_RATA' | 'PRIORITY' | 'AUCTION';

export interface Creditor {
  agent_id: string;
  amount_micros: number;
  priority: number;
  collateralized: boolean;
}

export interface Asset {
  asset_type: string;
  value_micros: number;
  liquid: boolean;
}

export interface Distribution {
  creditor_id: string;
  receives_micros: number;
  recovery_bps: number;
}

export interface DBP {
  dbp_version: string;
  default_id: string;
  defaulting_agent_id: string;
  declaration_type: DeclarationType;
  trigger: {
    type: TriggerType;
    reference_id: string;
    trigger_timestamp_ms: number;
  };
  obligations_snapshot: {
    total_owed_micros: number;
    creditors: Creditor[];
  };
  assets_snapshot: {
    total_value_micros: number;
    assets: Asset[];
  };
  recovery_rate_bps: number;
  liquidation_plan: {
    method: LiquidationMethod;
    distributions: Distribution[];
  };
  timestamp_ms: number;
  arbiter_agent_id: string;
  dbp_hash: string;
  signature_ed25519: string;
}

export interface MakeDBPParams {
  defaulting_agent_id: string;
  declaration_type: DeclarationType;
  trigger_type: TriggerType;
  trigger_reference_id: string;
  creditors: Creditor[];
  assets: Asset[];
  liquidation_method: LiquidationMethod;
  arbiter_agent_id: string;
  arbiter_private_key: string;
}

/**
 * Compute liquidation distributions based on method
 */
function computeDistributions(
  creditors: Creditor[],
  totalAssets: number,
  method: LiquidationMethod
): Distribution[] {
  // Handle empty creditors or zero owed
  if (creditors.length === 0) {
    return [];
  }

  const totalOwed = creditors.reduce((sum, c) => sum + c.amount_micros, 0);

  // Handle zero total owed
  if (totalOwed === 0) {
    return creditors.map(c => ({
      creditor_id: c.agent_id,
      receives_micros: 0,
      recovery_bps: 0
    }));
  }

  if (method === 'PRO_RATA') {
    return creditors.map(c => {
      const receives = Math.floor((c.amount_micros / totalOwed) * totalAssets);
      const recovery = c.amount_micros > 0 ? Math.floor((receives / c.amount_micros) * 10000) : 0;
      return {
        creditor_id: c.agent_id,
        receives_micros: receives,
        recovery_bps: recovery
      };
    });
  }

  if (method === 'PRIORITY') {
    const distributions: Distribution[] = [];
    let remaining = totalAssets;
    const sorted = [...creditors].sort((a, b) => a.priority - b.priority);

    for (const c of sorted) {
      const receives = Math.min(c.amount_micros, remaining);
      remaining -= receives;
      const recovery = c.amount_micros > 0 ? Math.floor((receives / c.amount_micros) * 10000) : 0;
      distributions.push({
        creditor_id: c.agent_id,
        receives_micros: receives,
        recovery_bps: recovery
      });
    }
    return distributions;
  }

  // AUCTION - same as PRO_RATA for now (auction logic external)
  return creditors.map(c => {
    const receives = Math.floor((c.amount_micros / totalOwed) * totalAssets);
    const recovery = c.amount_micros > 0 ? Math.floor((receives / c.amount_micros) * 10000) : 0;
    return {
      creditor_id: c.agent_id,
      receives_micros: receives,
      recovery_bps: recovery
    };
  });
}

/**
 * Create a signed DBP
 */
export async function make_dbp(params: MakeDBPParams): Promise<DBP> {
  const {
    defaulting_agent_id,
    declaration_type,
    trigger_type,
    trigger_reference_id,
    creditors,
    assets,
    liquidation_method,
    arbiter_agent_id,
    arbiter_private_key
  } = params;

  const now = Date.now();
  const totalOwed = creditors.reduce((sum, c) => sum + c.amount_micros, 0);
  const totalAssets = assets.reduce((sum, a) => sum + a.value_micros, 0);

  const distributions = computeDistributions(creditors, totalAssets, liquidation_method);
  const totalDistributed = distributions.reduce((sum, d) => sum + d.receives_micros, 0);
  const recovery_rate_bps = totalOwed > 0 ? Math.floor((totalDistributed / totalOwed) * 10000) : 0;

  // Create content for hashing (excludes default_id, dbp_hash, signature)
  const dbpContent = {
    dbp_version: '0.1',
    defaulting_agent_id,
    declaration_type,
    trigger: {
      type: trigger_type,
      reference_id: trigger_reference_id,
      trigger_timestamp_ms: now
    },
    obligations_snapshot: {
      total_owed_micros: totalOwed,
      creditors: [...creditors].sort((a, b) => a.agent_id.localeCompare(b.agent_id))
    },
    assets_snapshot: {
      total_value_micros: totalAssets,
      assets: [...assets].sort((a, b) => a.asset_type.localeCompare(b.asset_type))
    },
    recovery_rate_bps,
    liquidation_plan: {
      method: liquidation_method,
      distributions: distributions.sort((a, b) => a.creditor_id.localeCompare(b.creditor_id))
    },
    timestamp_ms: now,
    arbiter_agent_id
  };

  // Compute hash from content only
  const contentHash = hash(canonicalizeBytes(dbpContent));

  // Sign
  const signature = await sign(contentHash, arbiter_private_key);

  return {
    ...dbpContent,
    default_id: contentHash,
    dbp_hash: contentHash,
    signature_ed25519: signature
  };
}

/**
 * Extract hashable content from DBP (excludes default_id, dbp_hash, signature)
 */
function getDBPContent(dbp: DBP): object {
  const { default_id, dbp_hash, signature_ed25519, ...content } = dbp;
  return content;
}

/**
 * Verify DBP signature
 */
export async function verify_dbp(dbp: DBP): Promise<boolean> {
  const content = getDBPContent(dbp);
  const computedHash = hash(canonicalizeBytes(content));
  return verify(computedHash, dbp.signature_ed25519, dbp.arbiter_agent_id);
}

/**
 * Get DBP hash for referencing
 */
export function get_dbp_hash(dbp: DBP): string {
  const content = getDBPContent(dbp);
  return hash(canonicalizeBytes(content));
}

/**
 * Check if agent should trigger automatic default
 */
export function should_auto_default(
  runway_seconds: number,
  threshold_seconds: number = 0
): boolean {
  return runway_seconds < threshold_seconds;
}

/**
 * Cascade default calculation
 * Returns list of agents that would default if initial agent defaults
 */
export function calculate_cascade(
  initial_defaulter: string,
  agent_balances: Map<string, { runway_seconds: number; creditors: Creditor[] }>
): string[] {
  const cascaded: string[] = [initial_defaulter];
  const queue = [initial_defaulter];
  const processed = new Set<string>();

  while (queue.length > 0) {
    const defaulter = queue.shift()!;
    if (processed.has(defaulter)) continue;
    processed.add(defaulter);

    const defaulterData = agent_balances.get(defaulter);
    if (!defaulterData) continue;

    // Check each creditor's exposure
    for (const creditor of defaulterData.creditors) {
      if (processed.has(creditor.agent_id)) continue;

      const creditorData = agent_balances.get(creditor.agent_id);
      if (!creditorData) continue;

      // Simple: if creditor loses this exposure, check runway
      // In reality, would recalculate MBS properly
      const newRunway = creditorData.runway_seconds - (creditor.amount_micros / 1000000);

      if (newRunway < 0) {
        cascaded.push(creditor.agent_id);
        queue.push(creditor.agent_id);
      }
    }
  }

  return cascaded;
}

/**
 * Trigger default for an agent
 * Creates DBP with AUTOMATIC or INVOLUNTARY declaration
 */
export async function trigger_default(params: {
  defaulting_agent_id: string;
  trigger_type: TriggerType;
  trigger_reference_id: string;
  creditors: Creditor[];
  assets: Asset[];
  liquidation_method: LiquidationMethod;
  arbiter_agent_id: string;
  arbiter_private_key: string;
  declaration_type?: 'AUTOMATIC' | 'INVOLUNTARY';
}): Promise<DBP> {
  return make_dbp({
    ...params,
    declaration_type: params.declaration_type || 'AUTOMATIC'
  });
}

/**
 * Resolve default by creating final settlement DBP
 * Used after liquidation auction or pro-rata distribution
 */
export async function resolve_default(params: {
  original_dbp: DBP;
  final_distributions: Distribution[];
  arbiter_agent_id: string;
  arbiter_private_key: string;
}): Promise<DBP> {
  const { original_dbp, final_distributions, arbiter_agent_id, arbiter_private_key } = params;

  // Recompute recovery rates based on final distributions
  const totalDistributed = final_distributions.reduce((sum, d) => sum + d.receives_micros, 0);
  const totalOwed = original_dbp.obligations_snapshot.total_owed_micros;
  const recovery_rate_bps = totalOwed > 0 ? Math.floor((totalDistributed / totalOwed) * 10000) : 0;

  const resolvedContent = {
    dbp_version: '0.1',
    defaulting_agent_id: original_dbp.defaulting_agent_id,
    declaration_type: 'INVOLUNTARY' as DeclarationType,
    trigger: {
      type: 'TIMEOUT' as TriggerType,
      reference_id: original_dbp.default_id,
      trigger_timestamp_ms: Date.now()
    },
    obligations_snapshot: original_dbp.obligations_snapshot,
    assets_snapshot: original_dbp.assets_snapshot,
    recovery_rate_bps,
    liquidation_plan: {
      method: 'AUCTION' as LiquidationMethod,
      distributions: final_distributions.sort((a, b) => a.creditor_id.localeCompare(b.creditor_id))
    },
    timestamp_ms: Date.now(),
    arbiter_agent_id
  };

  const contentHash = hash(canonicalizeBytes(resolvedContent));
  const signature = await sign(contentHash, arbiter_private_key);

  return {
    ...resolvedContent,
    default_id: contentHash,
    dbp_hash: contentHash,
    signature_ed25519: signature
  };
}
